Windows: named pipe mounts

jstarks opened the pull request:
Current insider builds of Windows have support for mounting individual
named pipe servers from the host to the guest. This allows, for example,
exposing the docker engine's named pipe to a container.

This change allows the user to request such a mount via the normal bind
mount syntax in the CLI:

```
docker run -v \\.\pipe\docker_engine:\\.\pipe\docker_engine <args>
```

jstarks commented:
@jhowardmsft PTAL. One thing I didn't add was a version check; if you do this on an older Windows version it will silently ignore the pipe list. If you can suggest a proper version check, I can add one.

jhowardmsft commented on libcontainerd/client_windows.go:
```diff
@@ -229,20 +229,32 @@ func (clnt *client) createWindows(containerID string, checkpoint string, checkpo
+	var mds []hcsshim.MappedDir
+	var mps []hcsshim.MappedPipe
+	for _, mount := range spec.Mounts {
+		pipePrefix := `\\.\pipe\`
```
Constant?

jhowardmsft commented on libcontainerd/client_windows.go:
```diff
@@ -229,20 +229,32 @@ func (clnt *client) createWindows(containerID string, checkpoint string, checkpo
+				HostPath:          mount.Source,
+				ContainerPipeName: mount.Destination[len(pipePrefix):],
+			}
+			mps = append(mps, mp)
```
Could add a `continue` after this line, then can remove the `else` below and out-dent it.

jstarks replied:
Is that good go style? I thought since this was an either-or and not a special case + skip situation, the indentation was better.

jhowardmsft replied:
That's fair

jhowardmsft commented on libcontainerd/client_windows.go:
```diff
@@ -229,20 +229,32 @@ func (clnt *client) createWindows(containerID string, checkpoint string, checkpo
 		}
 	}
 	configuration.MappedDirectories = mds
+	configuration.MappedPipes = mps
```
For a version check, you could use `system.GetOSVersion()` and fail the attempt if build < ~16230. Here might be a good spot and check `mps.Length > 0`

jstarks replied:
I picked an older build number since the platform support has been available in insider builds for a little while.

jhowardmsft commented:
@johnstep FYI

jhowardmsft commented on libcontainerd/client_windows.go:
```diff
@@ -229,20 +230,35 @@ func (clnt *client) createWindows(containerID string, checkpoint string, checkpo
 		}
 	}
 	configuration.MappedDirectories = mds
+	if len(mps) > 0 && system.GetOSVersion().Build < 16210 { // replace with Win10 RS3 build number at RTM
```
If you mark this 'TODO @jhowardmsft LCOW Support:...', I'll have a better chance of remembering 😊

jstarks replied:
I think at RTM we just need to search the code base for GetOSVersion and reevaluate all the checks, like we did at Server 2016 release.

lowenna reviewed (approved):
LGTM

AkihiroSuda commented on volume/validate.go:
```diff
@@ -97,6 +97,15 @@ func validateMountConfig(mnt *mount.Mount, options ...func(*validateOpts)) error
+		if mnt.BindOptions != nil {
+			return &errMountConfig{mnt, errExtraField("BindOptions")}
+		}
+
```
`if mnt.ReadOnly { errExtraField("ReadOnly") }`?

AkihiroSuda commented:
- Please add integration test? (But probably `Skip()` on current windows CI?)
- No support for swarm-mode? (can be another PR?)

AkihiroSuda commented:
ping @jstarks

johnstep reviewed (approved):
LGTM

johnstep commented:
@jstarks, it looks good but will you address the feedback from @AkihiroSuda when you have a chance?

jstarks commented:
I've pushed a new change that adds an integration test (only runs against RS3) and adds the read-only check, as requested by @AkihiroSuda. Thanks for the suggestions!

Swarm will have to be a separate PR.

AkihiroSuda commented on integration-cli/requirements_test.go:
```diff
@@ -37,6 +37,12 @@ func DaemonIsWindows() bool {
 	return PlatformIs("windows")
 }
 
+func DaemonIsWindowsBuild(buildNumber int) func() bool {
```
`Is` -> `GreaterThan`?

jstarks replied:
How about AtLeast?

AkihiroSuda commented on integration-cli/docker_cli_run_windows_test.go:
```diff
@@ -0,0 +1,52 @@
+	containerPipeName := `\\.\pipe\docker-cli-test-pipe`
+	text := "hello from a pipe"
+	cmd := fmt.Sprintf("echo %s > %s", text, containerPipeName)
+	dockerCmd(c, "run", "-v", hostPipeName+":"+containerPipeName, testEnv.MinimalBaseImage(), "cmd", "/c", cmd)
```
IIUC, this requires newer CLI binary? (new test should be an API test)

jstarks replied:
I don't think this requires a CLI binary change since the bind mount parsing happens on the server side. When I ran this test locally I used an existing CLI.

Regardless, I didn't realize that new tests were at the API level, but I can change+move this if necessary.

AkihiroSuda replied:
Yes, please convert the test to API test?
https://forums.mobyproject.org/t/evolution-of-testing-moby/38

jstarks replied:
I don't see any existing API tests that launch a container using the API. They all use the CLI... E.g. `TestGetContainersAttachWebsocket`. Are there any samples I'm missing?

AkihiroSuda replied:
`TestContainersAPICreateMountsTmpfs` is one of API tests that is similar to your workload https://github.com/moby/moby/blob/9d87e6e0fb799d6ef3bb9a97bc523f8d343b5fb3/integration-cli/docker_api_containers_test.go#L1895

jstarks replied:
Ah, thanks! Updated. By going through the API directly I was able to find a bug, which is now fixed. I also added some unit tests in volume.

jstarks committed:
Windows: Add named pipe mount support

Current insider builds of Windows have support for mounting individual
named pipe servers from the host to the guest. This allows, for example,
exposing the docker engine's named pipe to a container.

This change allows the user to request such a mount via the normal bind
mount syntax in the CLI:

  docker run -v \\.\pipe\docker_engine:\\.\pipe\docker_engine <args>

Signed-off-by: John Starks <jostarks@microsoft.com>

simonferquel commented on volume/volume_test.go:
```diff
@@ -1,269 +1,275 @@
+			{`name:d:`, "local", mount.TypeVolume, `d:`, ``, `name`, "local", true, false},
+			{`name:d::ro`, "local", mount.TypeVolume, `d:`, ``, `name`, "local", false, false},
+			{`name:c:`, "", mount.TypeVolume, ``, ``, ``, "", true, true},
+			{`driver/name:c:`, "", mount.TypeVolume, ``, ``, ``, "", true, true},
```
add failing cases for \\.\pipe\foo:c:\data and c:\windows:\\.\pipe\foo

jstarks replied:
Good idea. Done.

jstarks commented on volume/validate.go:
```diff
@@ -30,10 +28,8 @@ func validateMountConfig(mnt *mount.Mount, options ...func(*validateOpts)) error
 		return &errMountConfig{mnt, err}
 	}
 
-	if !opts.skipAbsolutePathCheck {
```
This option is obsolete. I removed it.

jstarks commented on volume/volume_windows.go:
```diff
@@ -49,11 +44,13 @@ const (
 
 	// RXSource is the combined possibilities for a source
-	RXSource = `((?P<source>((` + RXHostDir + `)|(` + RXName + `))):)?`
+	RXSource = `((?P<source>((` + RXHostDir + `)|(` + RXName + `)|(` + RXPipe + `))):)?`
```
Named pipe semantics aren't quite the same as Unix socket semantics, unfortunately (e.g. for Docker's named pipe transport we simulate shutdown() by sending a zero-byte write). Given that, we're not going to support this kind of mapping at the platform level. However, we could add a relay mechanism to Docker to support this specific case with the semantics Docker expects.

I'm also thinking about whether we can add AF_UNIX support to Windows directly... then we could change Docker to bind to a Unix socket in Windows by default (as well as the named pipe, for compatibility reasons), and bind mounting this would be more natural and perhaps supportable in the platform. But that wouldn't be for LCOW v1.

jstarks replied:
Regarding two-pass regex, the second pass validation I've just now added to `validateMountConfig` so that it occurs even if you pass pre-parsed mounts via the API instead of via the raw (-v) field.

simonferquel replied:
Agreed, the validateMountConfig is a better place to validate that.

@dhiltgen looks like the named pipe / unix socket mapping we were counting on to make linux flavored ucp controller run on windows won't be so easy then... We'll still have to rely on tcp/tls to connect to the daemon

friism replied:
@simonferquel well right now the UCP agent is a windows container, and that can have the daemon pipe mounted in as a pipe.

jhowardmsft commented:
Just realized, HCSShim should probably be v0.5.28 rather than .26 (latest pre-`S`irupsen), or v0.6.1 if #34272 gets merged first (which is imminent).

jhowardmsft commented:
34272 is merged and HCSShim v0.6.1 is currently in moby/moby master. You can just drop that commit now, and all should be good.

simonferquel reviewed (approved):
LGTM

jstarks commented:
Fixed gofmt issue that was blocking janky.

jstarks commented:
Re-pushed to resolve merge conflicts

jhowardmsft commented:
Moving to merge when green CI

yongtang commented:
All Jenkins tests passed. Merging.
